server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -116,11 +127,13 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
         }
 
         checkIndexBlocks(clusterState, concreteIndices);
-
-        final Map<String, FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedMap(new HashMap<>());
         // This map is used to share the index response for indices which have the same index mapping hash to reduce the memory usage.
         final Map<String, FieldCapabilitiesIndexResponse> indexMappingHashToResponses = Collections.synchronizedMap(new HashMap<>());
         final Consumer<FieldCapabilitiesIndexResponse> handleIndexResponse = resp -> {
+            if (fieldCapTask.isCancelled() && fieldCapTask.notifyIfCancelled(listener)) {
```

Comment:
I added an extra `fieldCapTask.isCancelled()` to avoid synchronization in `notifyIfCancelled`.

Reply:
Can we improve this in `CancellableTask` itself?

```diff
diff --git a/server/src/main/java/org/elasticsearch/tasks/CancellableTask.java b/server/src/main/java/org/elasticsearch/tasks/CancellableTask.java
index 9010a9d99d3..fbb00cf6dbc 100644
--- a/server/src/main/java/org/elasticsearch/tasks/CancellableTask.java
+++ b/server/src/main/java/org/elasticsearch/tasks/CancellableTask.java
@@ -104,10 +104,10 @@ public class CancellableTask extends Task {
      */
     public final <T> boolean notifyIfCancelled(ActionListener<T> listener) {
         final TaskCancelledException taskCancelledException;
+        if (isCancelled() == false) {
+            return false;
+        }
         synchronized (this) {
-            if (isCancelled() == false) {
-                return false;
-            }
             taskCancelledException = getTaskCancelledException();
         } // NB releasing the mutex before notifying the listener
         listener.onFailure(taskCancelledException);
```

Reply:
yes, I pushed https://github.com/elastic/elasticsearch/pull/93716/commits/05e55d634c4be2162ff0ca7c4631620a3b1e9bf1.